srml/elections-phragmen/src/lib.rs

```diff
@@ -0,0 +1,957 @@
+
+	/// How much should be locked up in order to submit one's candidacy.
+	type CandidacyBond: Get<BalanceOf<Self>>;
+
+	/// How much should be locked up in order to be able to submit votes.
+	type VotingBond: Get<BalanceOf<Self>>;
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as PhragmenElection {
+		// ---- parameters
+		/// Number of seats to elect.
+		pub DesiredSeats get(desired_seats) config(): u32;
+		/// The total number of vote rounds that have happened, exclusive of the upcoming one.
+		pub ElectionRounds get(election_rounds): u32 = Zero::zero();
+		/// How long each seat is kept. This defined the next block number at which an election
+		/// round will happen.
+		pub TermDuration get(term_duration) config(): T::BlockNumber;
+
+		// ---- State
+		/// The current elected membership. Sorted.
+		pub Members get(members) config(): Vec<T::AccountId>;
+
+		/// Votes of a particular voter, with the round index of the votes.
+		pub VotesOf get(votes_of): linked_map T::AccountId => (Vec<T::AccountId>, u32);
+		/// Locked stake of a voter.
+		pub StakeOf get(stake_of): map T::AccountId => BalanceOf<T>;
+
+		/// The present candidate list. Always sorted.
+		pub Candidates get(candidates): Vec<T::AccountId>;
+		/// Current number of active candidates.
+		pub CandidateCount get(candidate_count): u32;
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+
+		const CandidacyBond: BalanceOf<T> = T::CandidacyBond::get();
+		const VotingBond: BalanceOf<T> = T::VotingBond::get();
+
+		/// Vote for a set of candidates for the upcoming round of election.
+		///
+		/// The `votes` should:
+		///   - not be empty.
+		///   - not be less than the number of candidates.
+		///
+		/// `who` cannot be an old voter who has not yet called [`remove_voter`].
+		///
+		/// Upon voting, the entire balance of `who` is locked and a bond amount is reserved.
+		#[weight = SimpleDispatchInfo::FixedNormal(750_000)]
+		fn vote(origin, votes: Vec<T::AccountId>) {
+			let who = ensure_signed(origin)?;
+
+			// TODO: use decode_len #3071.
+			let candidates = Self::candidates();
+			ensure!(!candidates.is_empty(), "number of candidates cannot be zero");
+			ensure!(candidates.len() >= votes.len(), "cannot vote more than candidates");
+
+			ensure!(!votes.is_empty(), "must vote for at least one candidate.");
+			ensure!(!Self::is_old_voter(&who), "must first remove self as voter and re-submit");
+
+			if !Self::is_current_voter(&who) {
+				// Amount to be locked up.
+				let locked_balance = T::Currency::total_balance(&who);
+
+				// lock and reserve
+				T::Currency::reserve(&who, T::VotingBond::get())
+					.map_err(|_| "voter can not pay voting bond")?;
+				T::Currency::set_lock(
+					MODULE_ID,
+					&who,
+					locked_balance,
+					T::BlockNumber::max_value(),
+					WithdrawReasons::except(WithdrawReason::TransactionPayment),
```

Comment:
this is fine for polkadot where 80% of tx fees go to the treasury, but if ever anyone did a chain where 100% transaction fees are given to the block author and that had arbitrary tipping in, then it would be a huge security flaw. you would simply lock your account, send a dumb transaction with a huge tip for a block that you authored, then collect the entirety of the "locked" (now unlocked) funds as the block author. the lock would remain in place, but the funds would not.

to make everything safer, you should provide a second argument `value: BalanceOf<T>` and that should be the amount that is locked and which is used for the purposes of voting. it should be interpreted as `value.min( T::Currency::total_balance(&who))` to avoid the need to faff around calculating how much is left in the account after tx fees if you just want to lock all of it.

Reply:
Fair enough and great catch -- then this needs updating in both election versions. We also do the same in the other election module. 

I interpret the suggested approach simply as: you as the voter decide how much of your stake is locked and taken into account when weighing the votes in either approval voting or phragmen. You are also responsible to keep some of your balance unlocked to be able to pay for some future txs.

Reply:
right (or you can lock all of it and transfer some extra in when you need to transact)